File: Management/Services/Files/CourseDifferences.cs
using LocalModels;

public static class CourseDifferences
{
  public static DeleteCourseChanges GetDeletedChanges(LocalCourse newCourse, LocalCourse oldCourse)
  {
    if (newCourse == oldCourse)
      return new DeleteCourseChanges();
    var moduleNamesNoLongerReferenced = oldCourse.Modules
      .Where(oldModule =>
        !newCourse.Modules.Any(newModule => newModule.Name == oldModule.Name)
      )
      .Select(oldModule => oldModule.Name)
      .ToList();

    var modulesWithDeletions = oldCourse.Modules
      .Where(oldModule =>
        !newCourse.Modules.Any(newModule => newModule.Equals(oldModule))
      )
      .Select(oldModule =>
      {
        var newModule = newCourse.Modules.FirstOrDefault(m => m.Name == oldModule.Name);
        if (newModule == null)
          return oldModule;

        var unreferencedAssignments = oldModule.Assignments.Where(oldAssignment =>
          !newModule.Assignments.Any(newAssignment => newAssignment.Name == oldAssignment.Name)
        );
        var unreferencedQuizzes = oldModule.Quizzes.Where(oldQuiz =>
          !newModule.Quizzes.Any(newQuiz => newQuiz.Name == oldQuiz.Name)
        );
        var unreferencedPages = oldModule.Pages.Where(oldPage =>
          !newModule.Pages.Any(newPage => newPage.Name == oldPage.Name)
        );

        return oldModule with
        {
          Assignments = unreferencedAssignments,
          Quizzes = unreferencedQuizzes,
          Pages = unreferencedPages,
        };
      })
      .ToList();

    return new DeleteCourseChanges
    {
      NamesOfModulesToDeleteCompletely = moduleNamesNoLongerReferenced,
      DeleteContentsOfModule = modulesWithDeletions,
    };
  }

  public static NewCourseChanges GetNewChanges(LocalCourse newCourse, LocalCourse oldCourse)
  {
    if (newCourse == oldCourse)
      return new NewCourseChanges();

    var differentModules = newCourse.Modules
      .Where(newModule =>
        !oldCourse.Modules.Any(oldModule => oldModule.Equals(newModule))
      )
      .Select(newModule =>
      {
        var oldModule = oldCourse.Modules.FirstOrDefault(m => m.Name == newModule.Name);
        if (oldModule == null)
          return newModule;

        var newAssignments = newModule.Assignments.Where(
          newAssignment => !oldModule.Assignments.Any(oldAssignment => newAssignment == oldAssignment)
        );
        var newQuizzes = newModule.Quizzes.Where(
          newQuiz => !oldModule.Quizzes.Any(oldQuiz => newQuiz == oldQuiz)
        );
        var newPages = newModule.Pages.Where(
          newPage => !oldModule.Pages.Any(oldPage => newPage == oldPage)
        );
        return newModule with
        {
          Assignments = newAssignments,
          Quizzes = newQuizzes,
          Pages = newPages,
        };
      })
      .ToArray();

    return new NewCourseChanges
    {
      Settings = newCourse.Settings,
      Modules = differentModules,
    };
  }

}

public record NewCourseChanges
{
  public IEnumerable<LocalModule> Modules { get; init; } = [];
  public LocalCourseSettings? Settings { get; init; }
}
public record DeleteCourseChanges
{
  public IEnumerable<string> NamesOfModulesToDeleteCompletely { get; init; } = [];
  public IEnumerable<LocalModule> DeleteContentsOfModule { get; init; } = [];
}

File: nextjs/src/services/fileStorage/utils/courseDifferences.ts
import { LocalCourse, LocalCourseSettings } from "@/models/local/localCourse";
import { LocalModule } from "@/models/local/localModules";

export const CourseDifferences = {
  getDeletedChanges(
    newCourse: LocalCourse,
    oldCourse: LocalCourse
  ): DeleteCourseChanges {
    if (newCourse === oldCourse) {
      const emptyDeletes: DeleteCourseChanges = {
        namesOfModulesToDeleteCompletely: [],
        deleteContentsOfModule: [],
      };
      return emptyDeletes;
    }

    const moduleNamesNoLongerReferenced = oldCourse.modules
      .filter(
        (oldModule) =>
          !newCourse.modules.some(
            (newModule) => newModule.name === oldModule.name
          )
      )
      .map((oldModule) => oldModule.name);

    const modulesWithDeletions = oldCourse.modules
      .filter(
        (oldModule) =>
          !newCourse.modules.some(
            (newModule) =>
              JSON.stringify(newModule) === JSON.stringify(oldModule)
          )
      )
      .map((oldModule) => {
        const newModule = newCourse.modules.find(
          (m) => m.name === oldModule.name
        );
        if (!newModule) return oldModule;

        const unreferencedAssignments = oldModule.assignments.filter(
          (oldAssignment) =>
            !newModule.assignments.some(
              (newAssignment) => newAssignment.name === oldAssignment.name
            )
        );
        const unreferencedQuizzes = oldModule.quizzes.filter(
          (oldQuiz) =>
            !newModule.quizzes.some((newQuiz) => newQuiz.name === oldQuiz.name)
        );
        const unreferencedPages = oldModule.pages.filter(
          (oldPage) =>
            !newModule.pages.some((newPage) => newPage.name === oldPage.name)
        );

        return {
          ...oldModule,
          assignments: unreferencedAssignments,
          quizzes: unreferencedQuizzes,
          pages: unreferencedPages,
        };
      });

    return {
      namesOfModulesToDeleteCompletely: moduleNamesNoLongerReferenced,
      deleteContentsOfModule: modulesWithDeletions,
    };
  },

  getNewChanges(
    newCourse: LocalCourse,
    oldCourse: LocalCourse
  ): NewCourseChanges {
    if (newCourse === oldCourse) {
      const emptyChanges: NewCourseChanges = {
        modules: [],
      };
      return emptyChanges;
    }

    const differentModules = newCourse.modules
      .filter(
        (newModule) =>
          !oldCourse.modules.some(
            (oldModule) =>
              JSON.stringify(oldModule) === JSON.stringify(newModule)
          )
      )
      .map((newModule) => {
        const oldModule = oldCourse.modules.find(
          (m) => m.name === newModule.name
        );
        if (!oldModule) return newModule;

        const newAssignments = newModule.assignments.filter(
          (newAssignment) =>
            !oldModule.assignments.some(
              (oldAssignment) =>
                JSON.stringify(newAssignment) === JSON.stringify(oldAssignment)
            )
        );
        const newQuizzes = newModule.quizzes.filter(
          (newQuiz) =>
            !oldModule.quizzes.some(
              (oldQuiz) => JSON.stringify(newQuiz) === JSON.stringify(oldQuiz)
            )
        );
        const newPages = newModule.pages.filter(
          (newPage) =>
            !oldModule.pages.some(
              (oldPage) => JSON.stringify(newPage) === JSON.stringify(oldPage)
            )
        );

        return {
          ...newModule,
          assignments: newAssignments,
          quizzes: newQuizzes,
          pages: newPages,
        };
      });

    return {
      settings: newCourse.settings,
      modules: differentModules,
    };
  },
};

export interface DeleteCourseChanges {
  namesOfModulesToDeleteCompletely: string[];
  deleteContentsOfModule: LocalModule[];
}

export interface NewCourseChanges {
  modules: LocalModule[];
  settings?: LocalCourseSettings;
}

// Default values for DeleteCourseChanges and NewCourseChanges
// export const createDeleteCourseChanges = (
//   init?: Partial<DeleteCourseChanges>
// ): DeleteCourseChanges => ({
//   namesOfModulesToDeleteCompletely: init?.namesOfModulesToDeleteCompletely ?? [],
//   deleteContentsOfModule: init?.deleteContentsOfModule ?? [],
// });

// export const createNewCourseChanges = (
//   init?: Partial<NewCourseChanges>
// ): NewCourseChanges => ({
//   modules: init?.modules ?? [],
//   settings: init?.settings,
// });
